### PR TITLE
feature: update cluster config

### PR DESCRIPTION
**Pull request by CadmusJiang**
reason: update tidb cluster config
way: 1.write new config to tiup's meta, and then call reload(cmd) to update involved components. 2. add udpate webhook check to prevent duplicate updates
test steps:
1.create Machine,TestResourceRequest,TestClusterTopology
2.update TestClusterTopology
3.ssh into control machine to updated components dir, check log
4.update twice continuously, the second time returns an error

Signed-off-by: Cadmus <CadmusJiang@gmail.com>

### CODE REVIEW

**Review by Hexilee — APPROVED** [review]
LGTM

File: controllers/testclustertopology_controller.go [hunk 1]
```diff
@@ -115,7 +115,52 @@ func (r *TestClusterTopologyReconciler) Reconcile(req ctrl.Request) (ctrl.Result
...
 		}
 	case naglfarv1.ClusterTopologyStateReady:
-		// DO NOTHING
+		// first create
+		if ct.Status.PreServerConfigs == nil && ct.Status.PreVersion == nil {
+			ct.Status.PreServerConfigs = ct.Spec.TiDBCluster.ServerConfigs.DeepCopy()
+			ct.Status.PreVersion = ct.Spec.TiDBCluster.Version.DeepCopy()
+			if err := r.Status().Update(ctx, &ct); err != nil {
+				log.Error(err, "unable to update TestClusterTopology")
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, err
```
Comment (mahjonp) [inline controllers/testclustertopology_controller.go:124]: You should select one of returning an error or re-queue.

File: controllers/testclustertopology_controller.go [hunk 2]
```diff
@@ -115,7 +115,52 @@ func (r *TestClusterTopologyReconciler) Reconcile(req ctrl.Request) (ctrl.Result
...
+			}
+			return ctrl.Result{}, nil
+		}
+		// no change
+		if !tiup.IsServerConfigModified(*ct.Status.PreServerConfigs, ct.Spec.TiDBCluster.ServerConfigs) {
+			log.Info("Cluster no change")
```
Comment (mahjonp) [inline controllers/testclustertopology_controller.go:130]: You should write a more clear log or event here.

File: controllers/testclustertopology_controller.go [hunk 3]
```diff
@@ -115,7 +115,52 @@ func (r *TestClusterTopologyReconciler) Reconcile(req ctrl.Request) (ctrl.Result
...
+			return ctrl.Result{}, nil
+		}
+		log.Info("Cluster is updating", "clusterName", ct.Name)
```
Comment (mahjonp) [inline controllers/testclustertopology_controller.go:133]: cluster is namespace-scoped, so logs should attach the ns name.

File: controllers/testclustertopology_controller.go [hunk 4]
```diff
@@ -115,7 +115,52 @@ func (r *TestClusterTopologyReconciler) Reconcile(req ctrl.Request) (ctrl.Result
...
+		ct.Status.State = naglfarv1.ClusterTopologyStateUpdating
+		if err := r.Status().Update(ctx, &ct); err != nil {
+			log.Error(err, "unable to update TestClusterTopology")
+			return ctrl.Result{}, err
+		}
```
Comment (mahjonp) [inline controllers/testclustertopology_controller.go:138]: It's better to `return ctrl.Result{}, nil` here IMO.

File: controllers/testclustertopology_controller.go [hunk 5]
```diff
@@ -115,7 +115,52 @@ func (r *TestClusterTopologyReconciler) Reconcile(req ctrl.Request) (ctrl.Result
...
+		requeue, err := r.updateTiDBCluster(ctx, &ct, &rr)
+		if err != nil {
+			r.Recorder.Event(&ct, "Warning", "Updating", err.Error())
+			return ctrl.Result{}, err
+		}
+		if requeue {
+			return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		//log.Info("Cluster is updated")
```
Comment (mahjonp) [inline controllers/testclustertopology_controller.go:155]: remove

File: pkg/tiup/deploy.go [hunk 6]
```diff
@@ -260,6 +298,20 @@ func (c *ClusterManager) UninstallCluster(clusterName string) error {
 	return nil
 }
 
+func (c *ClusterManager) diffTopology(pre naglfarv1.ServerConfigs, cur naglfarv1.ServerConfigs) []string {
```
Comment (mahjonp) [inline pkg/tiup/deploy.go:301]: bad name.

File: api/v1/testclustertopology_webhook.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *TestClusterTopology) ValidateUpdate(old runtime.Object) error {
+	testclustertopologylog.Info("validate update", "name", r.Name)
+	topology := old.(*TestClusterTopology)
+
+	if topology.Status.State == "updating" {
+		return errors.New("cluster is updating")
```
Comment (mahjonp) [inline api/v1/testclustertopology_webhook.go:66]: How to resolve the situation when you set an invalid configuration so the tiup fails to reload? You should give the user a path to revise the incorrect configration.

File: controllers/testclustertopology_controller.go (hunk 1 above)
Comment (CadmusJiang) [reply controllers/testclustertopology_controller.go:124]: done

Comment (CadmusJiang) [reply controllers/testclustertopology_controller.go:124]: change: ctrl.Result{},err

File: controllers/testclustertopology_controller.go (hunk 2 above)
Comment (CadmusJiang) [reply controllers/testclustertopology_controller.go:130]: change: remove

File: controllers/testclustertopology_controller.go (hunk 3 above)
Comment (CadmusJiang) [reply controllers/testclustertopology_controller.go:133]: done

File: controllers/testclustertopology_controller.go (hunk 4 above)
Comment (CadmusJiang) [reply controllers/testclustertopology_controller.go:138]: change: add return ctrl.Result{},err

File: controllers/testclustertopology_controller.go (hunk 5 above)
Comment (CadmusJiang) [reply controllers/testclustertopology_controller.go:155]: done

File: pkg/tiup/deploy.go (hunk 6 above)
Comment (CadmusJiang) [reply pkg/tiup/deploy.go:301]: change: diffSerConfigs

**Pull request merged**